src/mumble/pluginAPI/Plugin.h

```diff
@@ -0,0 +1,155 @@
+
+	/// Gets the description of the plugin
+	///
+	/// @param[out] descriptionBuffer The buffer into which the description shall be copied by this function
+	/// @param bufferSize The siue of the buffer - The length of the description (including the terminating zero-byte) must not exceed
+	/// 	this length.
+	PLUGIN_EXPORT void getDescription(char *descriptionBuffer, size_t bufferSize);
+
+	/// Provides the MumbleAPI struct to the plugin. This struct contains function pointers that can be used
+	/// to interact with the Mumble client. It is up to the plugin to store this struct somewhere if it wants to make use
+	/// of it at some point.
+	///
+	/// @param api The MumbleAPI struct
+	PLUGIN_EXPORT void setAPIFunctions(const struct MumbleAPI *api);
+
+
+
+	/// Sets the ID of this plugin. This is the ID Mumble will reference this plugin with and by which this plugin
+	/// can identify itself when communicating with Mumble.
+	///
+	/// @param id The ID for this plugin
+	PLUGIN_EXPORT void setPluginID(int32_t id);
+
+
+	// Parameters to functions below are yet to be determined
+
+	// -------- Callback functions -----------
+
+	/// Called when connecting to a server.
+	PLUGIN_EXPORT void onServerConnected();
+
+	/// Called when disconnecting from a server.
+	PLUGIN_EXPORT void onServerDisconnected();
+
+	/// Called whenever any user on the server enters a channel
+	/// This function will also be called when freshly connecting to a server as each user on that
+	/// server needs to be "added" to the respective channel as far as the local client is concerned.
+	///
+	/// @param userID The ID of the user this event has been triggered for
+	/// @param previousChannelID The ID of the chanel the user is coming from. Negative IDs indicate that there is no previous channel (e.g. the user
+	/// 	freshly connected to the server) or the channel isn't available because of any other reason.
+	/// @param newChannelID The ID of the channel the user has entered. If the ID is negative, the new channel could not be retrieved. This means
+	/// 	that the ID is invalid.
+	PLUGIN_EXPORT void onChannelEntered(uint32_t userID, int32_t previousChannelID, int32_t newChannelID);
+
+	/// Called whenever a user leaves a channel.
+	/// This function will also be called when the user disconnects from the server.
+	///
+	/// @param userID The ID of the user that left the channel
+	/// @param channelID The ID of the channel the user left. If the ID is negative, the channel could not be retrieved. This means that the ID is
+	/// 	invalid.
+	PLUGIN_EXPORT void onChannelExited(uint32_t userID, int32_t channelID);
+
+	/// Called when any user changes his/her talking state.
+	///
+	/// @param userID The ID of the user whose talking state has been changed
+	/// @param talkingState The new talking state the user has switched to. Currently the following talking states exist:
+	/// 	- <code>talkingState == 0</code> User is passive (not talking / not sending any audio)
```

Comment:
maybe just use a enum for this?

Reply:
I fear that when using an enum it won't be possible to adapt properly if someone was to introduce a new talking state into Mumble. In my eyes an enum suggests that no other cases have to be handled (e.g. that a default branch in a switch statement is unnecessary) whereas an integer suggests that there might be something new at some point.

Reply:
Yes that's correct, but what are the chances of a new talkingState to be added.

Wait, mumble has talking state for whispering/shouting?

Reply:
Rather slim, admittedly. Nonetheless I'd rather worry about it without it ever happening instead of it happening and me having to break API compatibility for something that could so easily be avoided.

I guess additionally adding a enum might be a compromise though? So that you have some useful enum constant names that actually tell you what they stand for in the code, but still leaving the parameter as an int in case a new state is being introduced.
After all enums are just glorified ints so implicit conversions should take care of the rest, shouldn't it

> Wait, mumble has talking state for whispering/shouting?

Apparently - From what I have seen in the code you can have whisperlists or something like this. Don't have a clue what shouting is though :man_shrugging:

Reply:
> and me having to break API compatibility for something that could so easily be avoided.
adding a new enum value doesn't break api compatibility, enum's are just passed as int

Reply:
Alright then - let's use an enum :+1: